bokehjs/src/lib/models/tools/actions/redo_tool.ts

```diff
@@ -11,7 +11,11 @@ export class RedoToolView extends ActionToolView {
   }
 
   doit(): void {
-    this.plot_view.state.redo()
+    const state = this.plot_view.state.redo()
+
+    if (state?.range != null) {
+      this._trigger_ranges_update()
```

Comment:
So now we have to keep track of this in every tool, each of which seem that have at least slightly different semantics. Are those events supposed to be triggered unconditionally, e.g. in this case if undo action didn't result in actual range change? In case of interactive updates, do we trigger at the beginning and end, or only at end, or it depends on a tool? I think this needs to be handled by plot canvas' range update logic instead of on each tool's individual basis.

Reply:
> Are those events supposed to be triggered unconditionally, e.g. in this case if undo action didn't result in actual range change? 
I don't think it should, which is why I did things as above. IMO a `RangesUpdate` should only trigger when ranges update. Additionally for "interactive" events that can also trigger LOD (e.g. scroll zoom) it should only trigger once, at the end, for all ranges.

Reply:
> I think this needs to be handled by plot canvas' range update logic instead of on each tool's individual basis.
I don't think this is possible. Some of the tools make many range updates, but we we don't want to trigger the event on all of them. Only the tool is guaranteed to know the exact details of when to trigger the event.